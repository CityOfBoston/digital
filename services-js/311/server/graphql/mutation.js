// @flow

import type { Context } from '.';
import type { CreateServiceRequestArgs } from '../services/Open311';

export const Schema = `
input CreateRequestAttribute {
  code: String!
  value: String!
}

type Mutation {
  createRequest (
    code: String!
    description: String!
    firstName: String
    lastName: String
    email: String
    phone: String
    address: String
    addressId: String
    mediaUrl: String
    location: LatLngIn
    attributes: [CreateRequestAttribute!]!
  ): Request!
}
`;

type CreateRequestArgs = {|
  code: string,
  description: string,
  firstName: ?string,
  lastName: ?string,
  email: ?string,
  phone: ?string,
  address: ?string,
  addressId: ?string,
  location: ?{
    lat: number,
    lng: number,
  },
  mediaUrl: ?string,
  attributes: { code: string, value: string }[],
|};

export const resolvers = {
  Mutation: {
    createRequest: (root: mixed, args: CreateRequestArgs, { open311 }: Context) => {
      const createArgs: CreateServiceRequestArgs = {
        service_code: args.code,
        description: args.description,
        first_name: args.firstName,
        last_name: args.lastName,
        email: args.email,
        phone: args.phone,
        media_url: args.mediaUrl,
        attributes: args.attributes,
      };

      if (args.address) {
        createArgs.address_string = args.address;
      }

      if (args.addressId) {
        createArgs.address_id = args.addressId;
      }

      if (args.location) {
        createArgs.lat = args.location.lat;
        createArgs.long = args.location.lng;
      }

      return open311.createRequest(createArgs).then((arr) => arr[0]);
    },
  },
};
