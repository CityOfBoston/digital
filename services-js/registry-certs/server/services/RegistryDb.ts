import { ConnectionPool, IResult } from 'mssql';
import {
  DatabaseConnectionOptions,
  createConnectionPool,
} from '@cityofboston/mssql-common';

import fs from 'fs';
import { promisify } from 'util';
import DataLoader from 'dataloader';
import Rollbar from 'rollbar';
import RegistryDbFake from './RegistryDbFake';

const readFile = promisify(fs.readFile);

export enum OrderType {
  DeathCertificate = 'DC',
  BirthCertificate = 'BC',
}

export interface DeathCertificate {
  CertificateID: number;
  'Registered Number': string;
  InOut: 'I' | '*' | '#';
  'Date of Death': string | null;
  'Decedent Name': string;
  'Last Name': string;
  'First Name': string;
  RegisteredYear: string;
  AgeOrDateOfBirth: string;
  Pending: number;
}

export interface DeathCertificateSearchResult extends DeathCertificate {
  ResultCount: number;
}

export interface BirthCertificate {
  CertificatesID: number;
  'Registered Number': string;
  InOut: boolean;
  'Date of Birth': string;
  'Certificate Name': string;
  'Last Name': string;
  'First Name': string;
  RegisteredYear: string;
  Impounded: boolean;
}

export interface AddOrderOptions {
  orderID: string;
  orderDate: Date;
  contactName: string;
  contactEmail: string;
  contactPhone: string;
  shippingName: string;
  shippingCompany: string;
  shippingAddr1: string;
  shippingAddr2: string;
  shippingCity: string;
  shippingState: string;
  shippingZIP: string;
  billingName: string;
  billingAddr1: string;
  billingAddr2: string;
  billingCity: string;
  billingState: string;
  billingZIP: string;
  billingLast4: string;
  serviceFee: number;
  idempotencyKey: string;
}

export interface AddOrderResult {
  OrderKey: number;
  ErrorMessage: string;
}

export interface FindOrderResult {
  OrderKey: number;
  OrderType: string;
  OrderDate: string;
  OrderStatus: string;
  ProcessDtTm: string | null;
  ContactName: string;
  ContactEmail: string;
  ContactPhone: string;
  ShippingName: string;
  ShippingCompany: string;
  ShippingAddr1: string;
  ShippingAddr2: string;
  ShippingCity: string;
  ShippingState: string;
  ShippingZIP: string;
  BillingName: string;
  BillingAddr1: string;
  BillingAddr2: string;
  BillingCity: string;
  BillingState: string;
  BillingZIP: string;
  CertificateIDs: string;
  CertificateQuantities: string;
  CertificateCost: number;
  ServiceFee: number;
  TotalCost: number;
}

export interface BirthCertificateRequestArgs {
  certificateLastName: string;
  certificateFirstName: string;
  alternativeSpellings: string;
  dateOfBirth: Date;
  parent1LastName: string;
  parent1FirstName: string;
  parent2LastName: string;
  parent2FirstName: string;
}

const MAX_ID_LOOKUP_LENGTH = 1000;

// Converts a list of key strings into an array of comma-separated strings,
// each no longer than maxLength.
//
// E.g.: ["12345", "67890", "abcde"] => ["12345,67890", "abcde"]
export function splitKeys(
  maxLength: number,
  keys: Array<string>
): Array<string> {
  const keyStrings: Array<string> = [];
  let currentKeyString = '';

  keys.forEach(key => {
    if (currentKeyString.length === 0) {
      currentKeyString = key;
    } else if (currentKeyString.length + key.length + 1 < maxLength) {
      currentKeyString = `${currentKeyString},${key}`;
    } else {
      keyStrings.push(currentKeyString);
      currentKeyString = key;
    }
  });

  if (currentKeyString.length > 0) {
    keyStrings.push(currentKeyString);
  }

  return keyStrings;
}

export default class RegistryDb {
  private pool: ConnectionPool;
  private lookupDeathCertificateLoader: DataLoader<
    string,
    DeathCertificate | null
  >;

  constructor(pool: ConnectionPool) {
    this.pool = pool;
    this.lookupDeathCertificateLoader = new DataLoader(keys =>
      this.lookupDeathCertificateLoaderFetch(keys)
    );
  }

  async searchDeathCertificates(
    name: string,
    page: number,
    pageSize: number,
    startYear: string | null | undefined,
    endYear: string | null | undefined
  ): Promise<Array<DeathCertificateSearchResult>> {
    const resp: IResult<DeathCertificateSearchResult> = (await this.pool
      .request()
      .input('searchFor', name)
      .input('pageNumber', page)
      .input('pageSize', pageSize)
      .input('sortBy', 'dateOfDeath')
      .input('startYear', startYear)
      .input('endYear', endYear)
      .execute('Registry.Death.sp_FindCertificatesWeb')) as any;

    const { recordset } = resp;

    if (!recordset) {
      throw new Error('Recordset for search came back empty');
    }

    return recordset;
  }

  async lookupDeathCertificate(id: string): Promise<DeathCertificate | null> {
    return this.lookupDeathCertificateLoader.load(id);
  }

  // "any" here is really DeathCertificate | null | Error
  private async lookupDeathCertificateLoaderFetch(
    keys: Array<string>
  ): Promise<Array<any>> {
    // The api can only take 1000 characters of keys at once. We probably won't
    // run into that issue but just in case we split up and parallelize.
    const keyStrings = splitKeys(MAX_ID_LOOKUP_LENGTH, keys);

    const idToOutputMap: {
      [key: string]: DeathCertificate | null | Error;
    } = {};

    const allResults: Array<Array<DeathCertificate>> = await Promise.all(
      keyStrings.map(async keyString => {
        try {
          const resp: IResult<DeathCertificate> = (await this.pool
            .request()
            .input('idList', keyString)
            .execute('Registry.Death.sp_GetCertificatesWeb')) as any;

          return resp.recordset;
        } catch (err) {
          keyString.split(',').forEach(id => (idToOutputMap[id] = err));
          return [];
        }
      })
    );

    allResults.forEach(results => {
      results.forEach((cert: DeathCertificate) => {
        idToOutputMap[cert.CertificateID.toString()] = cert;
      });
    });

    return keys.map(k => idToOutputMap[k]);
  }

  async addOrder(
    orderType: OrderType,
    {
      orderID,
      orderDate,
      contactName,
      contactEmail,
      contactPhone,
      shippingName,
      shippingCompany,
      shippingAddr1,
      shippingAddr2,
      shippingCity,
      shippingState,
      shippingZIP,
      billingName,
      billingAddr1,
      billingAddr2,
      billingCity,
      billingState,
      billingZIP,
      billingLast4,
      serviceFee,
      idempotencyKey,
    }: AddOrderOptions
  ): Promise<number> {
    const resp: IResult<AddOrderResult> = await this.pool
      .request()
      .input('orderID', orderID)
      .input('orderType', orderType)
      .input('orderDate', orderDate)
      .input('contactName', contactName)
      .input('contactEmail', contactEmail)
      .input('contactPhone', contactPhone)
      .input('shippingName', shippingName)
      .input('shippingCompany', shippingCompany)
      .input('shippingAddr1', shippingAddr1)
      .input('shippingAddr2', shippingAddr2)
      .input('shippingCity', shippingCity)
      .input('shippingState', shippingState)
      .input('shippingZIP', shippingZIP)
      .input('billingName', billingName)
      .input('billingAddr1', billingAddr1)
      .input('billingAddr2', billingAddr2)
      .input('billingCity', billingCity)
      .input('billingState', billingState)
      .input('billingZIP', billingZIP)
      .input('billingLast4', billingLast4)
      .input('serviceFee', `$${serviceFee.toFixed(2)}`)
      .input('idempotencyKey', idempotencyKey)
      .execute('Commerce.sp_AddOrder');

    const { recordset } = resp;

    if (!recordset || recordset.length === 0) {
      throw new Error('Recordset for creating an order came back empty');
    }

    const result = recordset[0];

    if (result.ErrorMessage) {
      throw new Error(result.ErrorMessage);
    }

    return result.OrderKey;
  }

  async addDeathCertificateItem(
    orderKey: number,
    certificateId: number,
    certificateName: string,
    quantity: number,
    certificateCost: number
  ): Promise<void> {
    const resp: IResult<Object> = await this.pool
      .request()
      .input('orderKey', orderKey)
      .input('orderType', OrderType.DeathCertificate)
      .input('certificateID', certificateId)
      .input('certificateName', certificateName)
      .input('quantity', quantity)
      .input('unitCost', `$${certificateCost.toFixed(2)}`)
      .execute('Commerce.sp_AddOrderItem');

    const { recordset } = resp;

    if (!recordset || recordset.length === 0) {
      throw new Error(
        `Could not add item to order ${orderKey}. Likely no certificate ID ${certificateId} in the database.`
      );
    }
  }

  async addBirthCertificateRequest(
    orderKey: number,
    {
      certificateFirstName,
      certificateLastName,
      alternativeSpellings,
      dateOfBirth,
      parent1FirstName,
      parent1LastName,
      parent2FirstName,
      parent2LastName,
    }: BirthCertificateRequestArgs,
    quantity: number,
    certificateCost: number
  ): Promise<number> {
    const resp: IResult<{
      RequestItemKey: number;
      ErrorMessage: string;
    }> = await this.pool
      .request()
      .input('orderKey', orderKey)
      .input('orderType', OrderType.BirthCertificate)
      .input('certificateLastName', certificateLastName)
      .input('certificateFirstName', certificateFirstName)
      .input('alternativeSpellings', alternativeSpellings)
      .input('dateOfBirth', dateOfBirth)
      .input('parent1LastName', parent1LastName)
      .input('parent1FirstName', parent1FirstName)
      .input('parent2LastName', parent2LastName)
      .input('parent2FirstName', parent2FirstName)
      .input('quantity', quantity)
      .input('unitCost', `$${certificateCost.toFixed(2)}`)
      .execute('Commerce.sp_AddBirthRequest');

    const { recordset } = resp;

    if (!recordset || recordset.length === 0) {
      throw new Error(`Could not add birth request to ${orderKey}.`);
    }

    if (recordset[0].ErrorMessage) {
      throw new Error(recordset[0].ErrorMessage);
    }

    return recordset[0].RequestItemKey;
  }

  async addPayment(
    orderKey: number,
    paymentDate: Date,
    transactionId: string,
    totalInDollars: number
  ): Promise<void> {
    const resp: IResult<Object> = await this.pool
      .request()
      .input('orderKey', orderKey)
      .input('paymentDate', paymentDate)
      .input('paymentDescription', '')
      .input('transactionID', transactionId)
      .input('paymentAmount', `$${totalInDollars.toFixed(2)}`)
      .execute('Commerce.sp_AddPayment');

    const { recordset } = resp;

    if (!recordset || recordset.length === 0) {
      throw new Error('Recordset for adding payment came back empty');
    }
  }

  async findOrder(orderId: string): Promise<FindOrderResult | null> {
    const resp: IResult<FindOrderResult> = await this.pool
      .request()
      .input('orderID', orderId)
      .execute('Commerce.sp_FindOrder');

    const { recordset } = resp;

    if (!recordset || recordset.length === 0) {
      return null;
    }

    return recordset[0];
  }

  async cancelOrder(orderKey: number, reason: string): Promise<void> {
    await this.pool
      .request()
      .input('orderKey', orderKey)
      .input('reason', reason)
      .execute('Commerce.sp_CancelOrder');
  }
}

export class RegistryDbFactory {
  protected pool: ConnectionPool;

  constructor(pool: ConnectionPool) {
    this.pool = pool;
  }

  registryDb() {
    return new RegistryDb(this.pool);
  }

  cleanup(): Promise<any> {
    return this.pool.close();
  }
}

export async function makeRegistryDbFactory(
  rollbar: Rollbar,
  connectionOptions: DatabaseConnectionOptions
): Promise<RegistryDbFactory> {
  const pool = await createConnectionPool(connectionOptions, err =>
    rollbar.error(err)
  );
  return new RegistryDbFactory(pool);
}

export async function makeFixtureRegistryDbFactory(
  fixtureName: string
): Promise<Required<RegistryDbFactory>> {
  const fixtureData = await readFile(fixtureName);
  const json = JSON.parse(fixtureData.toString('utf-8'));

  return {
    registryDb() {
      return new RegistryDbFake(json) as any;
    },

    cleanup() {
      return Promise.resolve(null);
    },
  };
}
