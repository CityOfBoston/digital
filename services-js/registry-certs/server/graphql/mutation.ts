/* eslint no-console: 0 */

import { Context } from '.';
import moment from 'moment';

import {
  Resolvers,
  ResolvableWith,
  Int,
} from '@cityofboston/graphql-typescript';

import { processChargeSucceeded } from '../stripe-events';

import {
  DEATH_CERTIFICATE_COST,
  calculateCreditCardCost,
  calculateDebitCardCost,
  BIRTH_CERTIFICATE_COST,
} from '../../lib/costs';

import makePaymentValidator from '../../lib/validators/PaymentValidator';
import makeShippingValidator from '../../lib/validators/ShippingValidator';
import { OrderType } from '../services/RegistryDb';

/** @graphql input */
interface DeathCertificateOrderItemInput {
  id: string;
  name: string;
  quantity: Int;
}

/**
 * This is the data registry needs to process a birth certificate request.
 *
 * @graphql input
 */
interface BirthCertificateOrderItemInput {
  firstName: string;
  lastName: string;
  alternateSpellings: string;
  birthDate: Date;
  parent1FirstName: string;
  parent1LastName: string;
  parent2FirstName: string;
  parent2LastName: string;
  relationship: string;
  quantity: Int;
}

/**
 * Either order or error will be non-null.
 */
interface OrderResult {
  order: SubmittedOrder | null;
  error: OrderError | null;
}

interface OrderError {
  message: string;
  cause: OrderErrorCause;
}

interface SubmittedOrder {
  id: string;
  contactEmail: string;
}

enum OrderErrorCause {
  /** Problem is a card error that the user should be able to correct. */
  USER_PAYMENT = 'USER_PAYMENT',
  INTERNAL = 'INTERNAL',
}

export interface Mutation extends ResolvableWith<{}> {
  submitDeathCertificateOrder(args: {
    contactName: string;
    contactEmail: string;
    contactPhone: string;

    shippingName: string;
    shippingCompanyName: string;
    shippingAddress1: string;
    shippingAddress2: string;
    shippingCity: string;
    shippingState: string;
    shippingZip: string;

    cardToken: string;
    cardLast4: string;

    cardholderName: string;
    billingAddress1: string;
    billingAddress2: string;
    billingCity: string;
    billingState: string;
    billingZip: string;

    items: DeathCertificateOrderItemInput[];
    idempotencyKey: string;
  }): OrderResult;

  submitBirthCertificateOrder(args: {
    contactName: string;
    contactEmail: string;
    contactPhone: string;

    shippingName: string;
    shippingCompanyName: string;
    shippingAddress1: string;
    shippingAddress2: string;
    shippingCity: string;
    shippingState: string;
    shippingZip: string;

    cardToken: string;
    cardLast4: string;

    cardholderName: string;
    billingAddress1: string;
    billingAddress2: string;
    billingCity: string;
    billingState: string;
    billingZip: string;

    item: BirthCertificateOrderItemInput;

    idempotencyKey: string;
  }): OrderResult;
}

const mutationResolvers: Resolvers<Mutation, Context> = {
  submitDeathCertificateOrder: async (
    _root,
    args,
    { rollbar, stripe, emails, registryDb }
  ): Promise<OrderResult> => {
    const {
      contactName,
      contactEmail,
      contactPhone,

      shippingName,
      shippingCompanyName,
      shippingAddress1,
      shippingAddress2,
      shippingCity,
      shippingState,
      shippingZip,

      cardToken,
      cardLast4,

      cardholderName,
      billingAddress1,
      billingAddress2,
      billingCity,
      billingState,
      billingZip,

      items,
      idempotencyKey,
    } = args;

    let totalQuantity = 0;
    items.forEach(({ quantity }) => {
      if (quantity <= 0) {
        throw new Error('Certificate quantity may not be less than 0');
      } else {
        totalQuantity += quantity;
      }
    });

    if (totalQuantity === 0) {
      throw new Error('Quantity of order is 0');
    }

    validateAddresses(args);

    // These are all in cents, to match Stripe
    const { total, serviceFee } = await calculateCostForToken(
      stripe,
      DEATH_CERTIFICATE_COST,
      cardToken,
      totalQuantity
    );

    const orderId = makeOrderId(OrderType.DeathCertificate);
    const orderDate = new Date();

    const orderKey = await registryDb.addOrder(OrderType.DeathCertificate, {
      orderID: orderId,
      orderDate,
      contactName,
      contactEmail,
      contactPhone,
      shippingName,
      shippingCompany: shippingCompanyName,
      shippingAddr1: shippingAddress1,
      shippingAddr2: shippingAddress2,
      shippingCity,
      shippingState,
      shippingZIP: shippingZip,
      billingName: cardholderName,
      billingAddr1: billingAddress1,
      billingAddr2: billingAddress2,
      billingCity,
      billingState,
      billingZIP: billingZip,
      billingLast4: cardLast4,
      serviceFee: serviceFee / 100,
      idempotencyKey,
    });

    await Promise.all(
      items.map(({ id, name, quantity }) =>
        registryDb.addDeathCertificateItem(
          orderKey,
          parseInt(id, 10),
          name,
          quantity,
          DEATH_CERTIFICATE_COST / 100
        )
      )
    );

    try {
      await makeStripeCharge(
        OrderType.DeathCertificate,
        { stripe, registryDb, rollbar, emails },
        {
          orderId,
          orderKey,
          quantity: totalQuantity,
          total,
          cardToken,
        }
      );
    } catch (e) {
      // These are user errors due to bad submissions (e.g. wrong CVV code)
      if (e.type === 'StripeCardError') {
        return {
          order: null,
          error: {
            message: e.message,
            cause: OrderErrorCause.USER_PAYMENT,
          },
        };
      } else {
        // This will cause it to get logged with Rollbar
        throw e;
      }
    }

    return { order: { id: orderId, contactEmail }, error: null };
  },

  submitBirthCertificateOrder: async (
    _root,
    args,
    { rollbar, stripe, emails, registryDb }
  ): Promise<OrderResult> => {
    const {
      contactName,
      contactEmail,
      contactPhone,

      shippingName,
      shippingCompanyName,
      shippingAddress1,
      shippingAddress2,
      shippingCity,
      shippingState,
      shippingZip,

      cardToken,
      cardLast4,

      cardholderName,
      billingAddress1,
      billingAddress2,
      billingCity,
      billingState,
      billingZip,

      item,

      idempotencyKey,
    } = args;

    if (item.quantity <= 0) {
      throw new Error('Certificate quantity may not be less than 0');
    }

    validateAddresses(args);

    // These are all in cents, to match Stripe
    const { total, serviceFee } = await calculateCostForToken(
      stripe,
      BIRTH_CERTIFICATE_COST,
      cardToken,
      item.quantity
    );

    const orderId = makeOrderId(OrderType.BirthCertificate);
    const orderDate = new Date();

    const orderKey = await registryDb.addOrder(OrderType.BirthCertificate, {
      orderID: orderId,
      orderDate,
      contactName,
      contactEmail,
      contactPhone,
      shippingName,
      shippingCompany: shippingCompanyName,
      shippingAddr1: shippingAddress1,
      shippingAddr2: shippingAddress2,
      shippingCity,
      shippingState,
      shippingZIP: shippingZip,
      billingName: cardholderName,
      billingAddr1: billingAddress1,
      billingAddr2: billingAddress2,
      billingCity,
      billingState,
      billingZIP: billingZip,
      billingLast4: cardLast4,
      serviceFee: serviceFee / 100,
      idempotencyKey,
    });

    await registryDb.addBirthCertificateRequest(
      orderKey,
      {
        certificateFirstName: item.firstName,
        certificateLastName: item.lastName,
        alternativeSpellings: item.alternateSpellings,
        dateOfBirth: item.birthDate,
        parent1FirstName: item.parent1FirstName,
        parent1LastName: item.parent1LastName,
        parent2FirstName: item.parent2FirstName,
        parent2LastName: item.parent2LastName,
      },
      item.quantity,
      BIRTH_CERTIFICATE_COST / 100
    );

    try {
      await makeStripeCharge(
        OrderType.BirthCertificate,
        { stripe, registryDb, rollbar, emails },
        {
          orderId,
          orderKey,
          quantity: item.quantity,
          total,
          cardToken,
        }
      );
    } catch (e) {
      // These are user errors due to bad submissions (e.g. wrong CVV code)
      if (e.type === 'StripeCardError') {
        return {
          order: null,
          error: {
            message: e.message,
            cause: OrderErrorCause.USER_PAYMENT,
          },
        };
      } else {
        // This will cause it to get logged with Rollbar
        throw e;
      }
    }

    return { order: { id: orderId, contactEmail }, error: null };
  },
};

/**
 * Throws an exception if the arguments do not match the shipping / payment
 * validators.
 */
function validateAddresses(args: {
  contactName: string;
  contactEmail: string;
  contactPhone: string;

  shippingName: string;
  shippingCompanyName: string;
  shippingAddress1: string;
  shippingAddress2: string;
  shippingCity: string;
  shippingState: string;
  shippingZip: string;

  cardToken: string;
  cardLast4: string;

  cardholderName: string;
  billingAddress1: string;
  billingAddress2: string;
  billingCity: string;
  billingState: string;
  billingZip: string;
}) {
  const {
    contactName,
    contactEmail,
    contactPhone,

    shippingName,
    shippingCompanyName,
    shippingAddress1,
    shippingAddress2,
    shippingCity,
    shippingState,
    shippingZip,

    cardholderName,
    billingAddress1,
    billingAddress2,
    billingCity,
    billingState,
    billingZip,
  } = args;

  const shippingValidator = makeShippingValidator({
    contactName,
    contactEmail,
    contactPhone,
    shippingName,
    shippingCompanyName,
    shippingAddress1,
    shippingAddress2,
    shippingCity,
    shippingState,
    shippingZip,
  });

  const paymentValidator = makePaymentValidator({
    cardholderName,
    billingAddressSameAsShippingAddress: false,
    billingAddress1,
    billingAddress2,
    billingCity,
    billingState,
    billingZip,
  });

  shippingValidator.check();
  paymentValidator.check();

  if (shippingValidator.fails() || paymentValidator.fails()) {
    const errors = {
      ...shippingValidator.errors.all(),
      ...paymentValidator.errors.all(),
    };
    const message = Object.keys(errors)
      .map(field => `${field}: ${errors[field].join(', ')}`)
      .join('\n');

    const err: any = new Error('Shipping validation errors');
    err.message = message;

    throw err;
  }
}

async function calculateCostForToken(
  stripe: any,
  eachCost: number,
  cardToken: string,
  quantity: number
) {
  // We have to look the token up again so we can figure out what fee
  // structure to use. We do *not* trust the client to send us this
  // information.
  const token = await stripe.tokens.retrieve(cardToken);

  return token.card.funding === 'credit'
    ? calculateCreditCardCost(eachCost, quantity)
    : calculateDebitCardCost(eachCost, quantity);
}

function makeOrderId(type: OrderType): string {
  let prefix: string;

  switch (type) {
    case OrderType.BirthCertificate:
      prefix = 'BC';
      break;
    case OrderType.DeathCertificate:
      prefix = 'DC';
      break;
    default:
      throw new Error('Unexpected OrderType: ' + type);
  }

  const datePart = moment().format('YYYYMM');

  // gives us a 6-digit number that doesn't start with 0
  const randomPart = 100000 + Math.floor(Math.abs(Math.random()) * 899999);

  return `RG-${prefix}${datePart}-${randomPart}`;
}

async function makeStripeCharge(
  type: OrderType,
  {
    stripe,
    registryDb,
    rollbar,
    emails,
  }: Pick<Context, 'stripe' | 'registryDb' | 'rollbar' | 'emails'>,
  {
    total,
    cardToken,
    orderId,
    orderKey,
    quantity,
  }: {
    total: number;
    cardToken: string;
    orderId: string;
    orderKey: number;
    quantity: number;
  }
): Promise<void> {
  let description: string;
  let unitPrice: number;
  let capture: boolean;

  switch (type) {
    case OrderType.DeathCertificate:
      description = 'Death certificates (Registry)';
      unitPrice = DEATH_CERTIFICATE_COST;
      capture = true;
      break;
    case OrderType.BirthCertificate:
      description = 'Birth certificates (Registry)';
      unitPrice = BIRTH_CERTIFICATE_COST;
      capture = false;
      break;
    default:
      throw new Error('Unknown OrderType: ' + type);
  }

  try {
    const charge = await stripe.charges.create({
      amount: total,
      currency: 'usd',
      source: cardToken,
      description,
      capture,
      statement_descriptor: 'CITYBOSTON*REG + FEE',
      metadata: {
        'webapp.name': 'registry-certs',
        'webapp.nodeEnv': process.env.NODE_ENV || 'development',
        'order.orderId': orderId,
        'order.orderKey': orderKey.toString(),
        'order.orderType': type,
        'order.source': 'registry',
        'order.quantity': quantity.toString(),
        'order.unitPrice': unitPrice.toString(),
      },
    });

    // The Stripe charge is the last thing in the request because if it
    // succeeds, we consider the order to have been completed successfully.
    // (This is certainly true from the customer’s perspective!)
    //
    // Sending the receipt email and marking the order as paid in the DB is
    // handled in response to a Stripe webhook for maximum reliability
    // (because Stripe will retry webhooks if they fail).

    // We can only get the Stripe callback when in production, so we fake it for dev.
    if (process.env.NODE_ENV === 'development') {
      await processChargeSucceeded({ stripe, emails, registryDb }, charge);
    }
  } catch (e) {
    try {
      await registryDb.cancelOrder(orderKey, 'Stripe charge create failed');
    } catch (e) {
      console.log('CANCEL ORDER FAILED');
      // Let Rollbar know, but still fail the mutation with the original
      // error.
      rollbar.error(e);
    }

    throw e;
  }
}

export const resolvers = {
  Mutation: mutationResolvers,
};
