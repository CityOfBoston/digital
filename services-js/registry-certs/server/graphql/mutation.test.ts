import { resolvers } from './mutation';
import RegistryDb from '../services/RegistryDb';

jest.mock('../services/RegistryDb');

const DEFAULT_ORDER = {
  contactName: 'Nancy Whitehead',
  contactEmail: 'nancy@mew.io',
  contactPhone: '(555) 123-4567',

  shippingName: 'Squirrel Girl',
  shippingCompanyName: 'Avengers Tower',
  shippingAddress1: '1 Avengers Pl.',
  shippingAddress2: '',
  shippingCity: 'New York',
  shippingState: 'NY',
  shippingZip: '10001',

  cardholderName: 'Doreen Green',
  cardToken: 'tok_test',
  cardLast4: '1234',

  billingAddress1: 'City University',
  billingAddress2: '',
  billingCity: 'New York',
  billingState: 'NY',
  billingZip: '10021',

  idempotencyKey: '1234abcd',
};

const DEFAULT_DEATH_ITEMS = [
  {
    id: '12345',
    name: '',
    quantity: 10,
  },
];

const DEFAULT_BIRTH_ITEM = {
  firstName: 'Danielle',
  lastName: 'Cage',
  relationship: 'parent',
  alternateSpellings: '',
  birthDate: new Date('01 March 2006 00:00:00 GMT'),
  parent1FirstName: 'Jessica',
  parent1LastName: 'Jones',
  parent2FirstName: 'Luke',
  parent2LastName: 'Cage',
  quantity: 10,
};

describe('Mutation resolvers', () => {
  let registryDb: jest.Mocked<RegistryDb>;

  beforeEach(() => {
    registryDb = new RegistryDb(null as any) as any;
  });

  describe('submitDeathCertificateOrder', () => {
    it('throws if a validator fails', async () => {
      await expect(
        resolvers.Mutation.submitDeathCertificateOrder(
          {},
          {
            ...DEFAULT_ORDER,

            items: DEFAULT_DEATH_ITEMS,
            contactName: '',
          },
          {} as any
        )
      ).rejects.toMatchSnapshot();
    });

    it('throws if there are no items', async () => {
      await expect(
        resolvers.Mutation.submitDeathCertificateOrder(
          {},
          {
            ...DEFAULT_ORDER,
            items: [],
          },
          {} as any
        )
      ).rejects.toMatchSnapshot();
    });

    it('sends a charge to Stripe', async () => {
      const tokensRetrieve = jest.fn();
      const chargesCreate = jest.fn();

      const stripe: any = {
        charges: {
          create: chargesCreate,
        },
        tokens: {
          retrieve: tokensRetrieve,
        },
      };

      const emails = {
        sendReceiptEmail: jest.fn(),
      };

      registryDb.addOrder.mockReturnValue(Promise.resolve(25));

      tokensRetrieve.mockReturnValue(
        Promise.resolve({ card: { funding: 'credit' } })
      );
      chargesCreate.mockReturnValue(Promise.resolve({ id: 'ch_12345' }));

      await resolvers.Mutation.submitDeathCertificateOrder(
        {},
        { ...DEFAULT_ORDER, items: DEFAULT_DEATH_ITEMS },
        {
          stripe,
          registryDb,
          emails,
        } as any
      );

      expect(chargesCreate).toHaveBeenCalledWith({
        amount: 14326,
        currency: 'usd',
        source: 'tok_test',
        capture: true,
        description: 'Death certificates (Registry)',
        statement_descriptor: 'CITYBOSTON*REG + FEE',
        metadata: expect.objectContaining({
          'webapp.name': 'registry-certs',
          'webapp.nodeEnv': 'test',
          'order.orderId': expect.any(String),
          'order.orderKey': '25',
          'order.orderType': 'DC',
          'order.quantity': '10',
          'order.unitPrice': '1400',
        }),
      });
    });
  });

  describe('submitBirthCertificateOrder', () => {
    it('throws if a validator fails', async () => {
      await expect(
        resolvers.Mutation.submitBirthCertificateOrder(
          {},
          {
            ...DEFAULT_ORDER,
            item: DEFAULT_BIRTH_ITEM,
            contactName: '',
          },
          { registryDb } as any
        )
      ).rejects.toMatchSnapshot();
    });

    it('sends a charge to Stripe', async () => {
      const tokensRetrieve = jest.fn();
      const chargesCreate = jest.fn();

      const stripe: any = {
        charges: {
          create: chargesCreate,
        },
        tokens: {
          retrieve: tokensRetrieve,
        },
      };

      const emails = {
        sendReceiptEmail: jest.fn(),
      };

      registryDb.addOrder.mockReturnValue(Promise.resolve(25));

      tokensRetrieve.mockReturnValue(
        Promise.resolve({ card: { funding: 'credit' } })
      );
      chargesCreate.mockReturnValue(Promise.resolve({ id: 'ch_12345' }));

      await resolvers.Mutation.submitBirthCertificateOrder(
        {},
        {
          ...DEFAULT_ORDER,
          item: {
            firstName: 'Doreen',
            lastName: 'Green',
            alternateSpellings: 'Squirrel Girl',
            birthDate: new Date('1997-07-01T00:00:00Z'),
            parent1FirstName: 'Maureen',
            parent1LastName: 'Green',
            parent2FirstName: 'Dor',
            parent2LastName: 'Green',
            quantity: 10,
            relationship: 'sidekick',
          },
        },

        {
          stripe,
          registryDb,
          emails,
        } as any
      );

      expect(chargesCreate).toHaveBeenCalledWith({
        amount: 14326,
        currency: 'usd',
        source: 'tok_test',
        capture: false,
        description: 'Birth certificates (Registry)',
        statement_descriptor: 'CITYBOSTON*REG + FEE',
        metadata: expect.objectContaining({
          'webapp.name': 'registry-certs',
          'webapp.nodeEnv': 'test',
          'order.orderId': expect.any(String),
          'order.orderKey': '25',
          'order.quantity': '10',
          'order.source': 'registry',
          'order.orderType': 'BC',
          'order.unitPrice': '1400',
        }),
      });
    });
  });
});
