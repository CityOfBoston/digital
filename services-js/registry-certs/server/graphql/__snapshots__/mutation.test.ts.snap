// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation resolvers submitBirthCertificateOrder throws if a validator fails 1`] = `[Error: contactName: The full name field is required.]`;

exports[`Mutation resolvers submitDeathCertificateOrder throws if a validator fails 1`] = `[Error: contactName: The full name field is required.]`;

exports[`Mutation resolvers submitDeathCertificateOrder throws if there are no items 1`] = `[Error: Quantity of order is 0]`;
